File: sdks/java/src/main/java/dev/atap/sdk/api/ApprovalApi.java
package dev.atap.sdk.api;

import com.fasterxml.jackson.databind.ObjectMapper;
import dev.atap.sdk.ATAPClient;
import dev.atap.sdk.model.Approval;
import dev.atap.sdk.model.ApprovalSubject;

import java.util.*;

/**
 * Create, respond to, list, and revoke approvals.
 */
public class ApprovalApi {

    private final ATAPClient client;

    public ApprovalApi(ATAPClient client) {
        this.client = client;
    }

    /**
     * Create an approval request. Requires atap:send scope.
     *
     * @param fromDid requester DID
     * @param toDid   approver DID (or org DID for fan-out)
     * @param subject the approval subject
     * @param via     optional mediating system DID
     * @return created Approval
     */
    public Approval create(String fromDid, String toDid, ApprovalSubject subject, String via) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("from", fromDid);
        body.put("to", toDid);

        Map<String, Object> subjectMap = new LinkedHashMap<>();
        subjectMap.put("type", subject.getType());
        subjectMap.put("label", subject.getLabel());
        subjectMap.put("payload", subject.getPayload() != null ? subject.getPayload() : Collections.emptyMap());
        body.put("subject", subjectMap);

        if (via != null && !via.isEmpty()) {
            body.put("via", via);
        }

        Map<String, Object> data = client.authedRequest("POST", "/v1/approvals", body, null, null, null);
        return mapToApproval(data);
    }

    public Approval create(String fromDid, String toDid, ApprovalSubject subject) {
        return create(fromDid, toDid, subject, null);
    }

    /**
     * Respond to an approval (approve). Requires atap:send scope.
     */
    public Approval respond(String approvalId, String signature) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("signature", signature);

        Map<String, Object> data = client.authedRequest("POST",
                "/v1/approvals/" + approvalId + "/respond", body, null, null, null);
        return mapToApproval(data);
    }

    /**
     * List approvals addressed to the authenticated entity. Requires atap:inbox scope.
     */
    @SuppressWarnings("unchecked")
    public List<Approval> list() {
        Map<String, Object> data = client.authedRequest("GET", "/v1/approvals", null, null, null, null);
        List<Map<String, Object>> items;
        if (data.containsKey("approvals")) {
            items = (List<Map<String, Object>>) data.get("approvals");
        } else if (data.containsKey("items")) {
            items = (List<Map<String, Object>>) data.get("items");
        } else {
            items = Collections.emptyList();
        }
        List<Approval> result = new ArrayList<>();
        for (Map<String, Object> item : items) {
            result.add(mapToApproval(item));
        }
        return result;
    }

    /**
     * Revoke an approval. Requires atap:revoke scope.
     */
    public Approval revoke(String approvalId) {
        Map<String, Object> data = client.authedRequest("DELETE",
                "/v1/approvals/" + approvalId, null, null, null, null);
        return mapToApproval(data);
    }

    @SuppressWarnings("unchecked")
    static Approval mapToApproval(Map<String, Object> data) {
        ObjectMapper mapper = new ObjectMapper();
        return mapper.convertValue(data, Approval.class);
    }
}

File: sdks/php/src/Api/ApprovalApi.php
<?php

declare(strict_types=1);

namespace Atap\Sdk\Api;

use Atap\Sdk\ATAPClient;
use Atap\Sdk\Model\Approval;
use Atap\Sdk\Model\ApprovalSubject;

/**
 * Create, respond to, list, and revoke approvals.
 */
class ApprovalApi
{
    public function __construct(
        private readonly ATAPClient $client,
    ) {
    }

    /**
     * Create an approval request. Requires atap:send scope.
     *
     * @param string $fromDid Requester DID
     * @param string $toDid Approver DID (or org DID for fan-out)
     * @param ApprovalSubject $subject The approval subject
     * @param string|null $via Optional mediating system DID
     */
    public function create(
        string $fromDid,
        string $toDid,
        ApprovalSubject $subject,
        ?string $via = null,
    ): Approval {
        $body = [
            'from' => $fromDid,
            'to' => $toDid,
            'subject' => $subject->toArray(),
        ];
        if ($via !== null) {
            $body['via'] = $via;
        }

        $data = $this->client->authedRequest('POST', '/v1/approvals', jsonBody: $body);

        return Approval::fromArray($data);
    }

    /**
     * Respond to an approval (approve). Requires atap:send scope.
     *
     * @param string $approvalId The approval ID (apr_...)
     * @param string $signature JWS signature from the approver
     */
    public function respond(string $approvalId, string $signature): Approval
    {
        $data = $this->client->authedRequest(
            'POST',
            "/v1/approvals/{$approvalId}/respond",
            jsonBody: ['signature' => $signature],
        );

        return Approval::fromArray($data);
    }

    /**
     * List approvals addressed to the authenticated entity. Requires atap:inbox scope.
     *
     * @return Approval[]
     */
    public function list(): array
    {
        $data = $this->client->authedRequest('GET', '/v1/approvals');

        $items = $data;
        if (!isset($data[0])) {
            $items = $data['approvals'] ?? $data['items'] ?? [];
        }

        return array_map(fn(array $a) => Approval::fromArray($a), $items);
    }

    /**
     * Revoke an approval. Requires atap:revoke scope.
     */
    public function revoke(string $approvalId): Approval
    {
        $data = $this->client->authedRequest('DELETE', "/v1/approvals/{$approvalId}");

        return Approval::fromArray($data);
    }
}
